src/internal/pfsdb/commits.go

```diff
@@ -0,0 +1,956 @@
+		CommitSetID:    commitInfo.Commit.Id,
+		RepoName:       commitInfo.Commit.Repo.Name,
+		RepoType:       commitInfo.Commit.Repo.Type,
+		ProjectName:    commitInfo.Commit.Repo.Project.Name,
+		BranchName:     sql.NullString{String: commitInfo.Commit.Branch.Name, Valid: true},
+		Origin:         commitInfo.Origin.Kind.String(),
+		StartTime:      sanitizeTimestamppb(commitInfo.Started),
+		FinishingTime:  sanitizeTimestamppb(commitInfo.Finishing),
+		FinishedTime:   sanitizeTimestamppb(commitInfo.Finished),
+		Description:    commitInfo.Description,
+		CompactingTime: durationpbToBigInt(commitInfo.Details.CompactingTime),
+		ValidatingTime: durationpbToBigInt(commitInfo.Details.ValidatingTime),
+		Size:           commitInfo.Details.SizeBytes,
+		Error:          commitInfo.Error,
+		ChildCommits:   commitInfo.ChildCommits,
+		ParentCommit:   commitInfo.ParentCommit,
+	}
+	query := updateCommit
+	_, err := tx.NamedExecContext(ctx, query, update)
+	if err != nil {
+		return errors.Wrap(err, "exec update commitInfo")
+	}
+	if !opt.SkipParent {
+		_, err = tx.ExecContext(ctx, "DELETE FROM pfs.commit_ancestry WHERE to_id=$1;", id)
+		if err != nil {
+			return errors.Wrap(err, "delete commit parent")
+		}
+		if commitInfo.ParentCommit != nil {
+			if err := PutCommitParent(ctx, tx, commitInfo.ParentCommit, id); err != nil {
+				return errors.Wrap(err, "linking parent")
+			}
+		}
+	}
+	if opt.SkipChildren {
+		return nil
+	}
+	_, err = tx.ExecContext(ctx, "DELETE FROM pfs.commit_ancestry WHERE from_id=$1;", id)
+	if err != nil {
+		return errors.Wrap(err, "delete commit children")
+	}
+	if len(commitInfo.ChildCommits) != 0 {
+		if err := PutCommitChildren(ctx, tx, id, commitInfo.ChildCommits); err != nil {
+			return errors.Wrap(err, "linking children")
+		}
+	}
+	return nil
+}
+
+// CommitPair is an (id, commitInfo) tuple returned by the commit iterator.
+type CommitPair struct {
+	ID         CommitID
+	CommitInfo *pfs.CommitInfo
+}
+
+// these dropped global variable instantiations forces the compiler to check whether CommitIterator and CommitIteratorTx implements stream.Iterator.
+var _ stream.Iterator[CommitPair] = &CommitIterator{}
+var _ stream.Iterator[CommitPair] = &CommitIteratorTx{}
+
+// CommitIterator batches a page of commitRow entries. (id, entry) tuples can be retrieved using iter.Next().
+type CommitIterator struct {
+	*commitIterator
+	db *pachsql.DB
+}
+
+// CommitIteratorTx is like CommitIterator but retrieves all pages within a single transaction.
+type CommitIteratorTx struct {
+	*commitIterator
+	tx *pachsql.Tx
+}
+
+type commitIterator struct {
+	rev     bool
+	limit   int
+	offset  int
+	commits []commitRow
+	index   int
+	filter  CommitListFilter
+}
+
+// Next advances the iterator by one row. It returns a stream.EOS when there are no more entries.
+func (iter *CommitIterator) Next(ctx context.Context, dst *CommitPair) error {
+	return errors.Wrap(dbutil.WithTx(ctx, iter.db, func(ctx context.Context, tx *pachsql.Tx) error {
```

Comment:
We shouldn't be opening a new transaction on every call to Next.

Reply:
Inside `Next` we should:
1.  Check if we have any commit pairs buffered.
2. If we do, emit, one of those, and remove it from the buffer
3. If we are out, open a transaction to get another page
4. Update the buffer, and also the query parameters for the next time.

Reply:
good catch, yeah this was on oversight on my part. we should only open a new transaction if we need to retrieve a new page.